numpy/ma/core.py

```diff
@@ -4287,22 +4298,40 @@ def count(self, axis=None):
         array([3, 0])
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         m = self._mask
-        s = self.shape
         if m is nomask:
+            # see _count_reduce_items in _methods.py
+            axes = axis
             if axis is None:
-                return self.size
+                axes = tuple(range(self.ndim))
+            if not isinstance(axes, tuple):
+                axes = (axis,)
+
+            items = 1
+            for ax in axes:
+                items *= self.shape[ax]
+
+            if self.shape is () or axis is None:
+                return items
+            elif kwargs.get('keepdims', False):
+                out_dims = list(self.shape)
+                for a in axes:
+                    out_dims[a] = 1
             else:
-                n = s[axis]
-                t = list(s)
-                del t[axis]
-                return np.full(t, n, dtype=np.intp)
-        n1 = np.size(m, axis)
-        n2 = np.sum(m, axis=axis, dtype=np.intp)
-        if axis is None:
-            return (n1 - n2)
-        else:
-            return narray(n1 - n2)
+                out_dims = [d for n,d in enumerate(self.shape) if n not in axes]
+            return np.asarray(np.intp(items)*ones(out_dims, dtype=np.intp))
+
+        # take care of the masked singleton
+        if self is masked:
+            return 0
+
+        # use data's class (eg so if it's a matrix, keeps dims)
```

Comment:
You're probably right. If my goal is that `count` behaves the same as `sum` with respect to keepdims, it seems like I always have so use `sum` to compute the count.

(Maybe I can get away with making the block above only apply to ndarrays though, as an optimization)